opensearch-operator/pkg/reconcilers/securityconfig.go

```diff
@@ -0,0 +1,213 @@
+		return ctrl.Result{}, nil
+	}
+	namespace := r.instance.Namespace
+	clusterName := r.instance.Name
+	jobName := clusterName + "-securityconfig-update"
+	configSecretName := r.instance.Spec.Security.Config.SecurityconfigSecret.Name
+	adminCertName := r.determineAdminSecret()
+
+	if adminCertName == "" {
+		r.logger.Info("Cluster is running with demo certificates. Skipping securityconfig")
+		return ctrl.Result{}, nil
+	}
+
+	// Wait for secret to be available
+	configSecret := corev1.Secret{}
+	if err := r.Get(r.ctx, client.ObjectKey{Name: configSecretName, Namespace: namespace}, &configSecret); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Info(fmt.Sprintf("Waiting for secret '%s' that contains the securityconfig to be created", configSecretName))
+			return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, nil
+		}
+		return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, err
+	}
+
+	// Calculate checksum and check for changes
+	checksum, err := checksum(configSecret.Data)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	job := batchv1.Job{}
+	if err := r.Get(r.ctx, client.ObjectKey{Name: jobName, Namespace: namespace}, &job); err == nil {
+		value, exists := job.ObjectMeta.Annotations[checksumAnnotation]
+		if exists && value == checksum {
+			// Nothing to do, current securityconfig already applied
+			return ctrl.Result{}, nil
+		}
+		// Delete old job
+		r.logger.Info("Deleting old update job")
+		opts := client.DeleteOptions{}
+		// Add this so pods of the job are deleted as well, otherwise they would remain as orphaned pods
+		client.PropagationPolicy(metav1.DeletePropagationForeground).ApplyToDelete(&opts)
+		err = r.Delete(r.ctx, &job, &opts)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		// Make sure job is completely deleted (when r.Delete returns deletion sometimes is not yet complete)
+		_, err = r.ReconcileResource(&job, reconciler.StateAbsent)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	r.logger.Info("Starting securityconfig update job")
+	job = r.job(jobName, namespace, checksum, adminCertName)
+	if err := ctrl.SetControllerReference(r.instance, &job, r.Client.Scheme()); err != nil {
+		return ctrl.Result{}, err
+	}
+	_, err = r.ReconcileResource(&job, reconciler.StateCreated)
+	return ctrl.Result{}, err
+}
+
+func checksum(data map[string][]byte) (string, error) {
+	hash := sha1.New()
+	keys := make([]string, 0, len(data))
+	for k := range data {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		_, err := hash.Write([]byte(key))
+		if err != nil {
+			return "", err
+		}
+		value := data[key]
+		_, err = hash.Write(value)
+		if err != nil {
+			return "", err
+		}
+	}
+	return base64.StdEncoding.EncodeToString(hash.Sum(nil)), nil
+}
+
+func (r *SecurityconfigReconciler) job(jobName string, namespace string, checksum string, adminCertName string) batchv1.Job {
```

Comment:
Didn’t understand whats the job purpose.

Reply:
The job will run the internal security script to update the credentials in the cluster.  Internally the opensearch docker image runs this script on start up, it's cleaner to have it running in a job to update the API when it gets changed.